src/components/MAPI/apps/utils.ts

```diff
@@ -0,0 +1,270 @@
+  try {
+    const cr = await corev1.getConfigMap(client, auth, name, namespace);
+    cr.data = {
+      values: contents,
+    };
+
+    return corev1.updateConfigMap(client, auth, cr);
+  } catch (err: unknown) {
+    if (
+      !metav1.isStatusError(
+        (err as GenericResponse).data,
+        metav1.K8sStatusErrorReasons.NotFound
+      )
+    ) {
+      return Promise.reject(err);
+    }
+  }
+
+  if (contents.length < 1) return null;
+
+  const cr: corev1.IConfigMap = {
+    apiVersion: 'v1',
+    kind: 'ConfigMap',
+    metadata: {
+      name,
+      namespace,
+    },
+    data: {
+      values: contents,
+    },
+  };
+
+  return corev1.createConfigMap(client, auth, cr);
+}
+
+/**
+ * Make sure that an app has a `Secret` with the given contents.
+ *
+ * If a `Secret` already exists, then its contents will be updated.
+ * If it doesn't exist, it will be created. Keep in mind that
+ * if an empty string is given, the `Secret` will not be created.
+ * @param client
+ * @param auth
+ * @param namespace
+ * @param name
+ * @param contents
+ */
+async function ensureAppUserSecret(
+  client: IHttpClient,
+  auth: IOAuth2Provider,
+  namespace: string,
+  name: string,
+  contents: string
+): Promise<corev1.ISecret | null> {
+  try {
+    const cr = await corev1.getSecret(client, auth, name, namespace);
+    cr.stringData = {
+      values: contents,
+    };
+
+    return corev1.updateSecret(client, auth, cr);
+  } catch (err: unknown) {
+    if (
+      !metav1.isStatusError(
+        (err as GenericResponse).data,
+        metav1.K8sStatusErrorReasons.NotFound
+      )
+    ) {
+      return Promise.reject(err);
+    }
+  }
+
+  if (contents.length < 1) return null;
+
+  const cr: corev1.ISecret = {
+    apiVersion: 'v1',
+    kind: 'Secret',
+    metadata: {
+      name,
+      namespace,
+    },
+    stringData: {
+      values: contents,
+    },
+  };
+
+  return corev1.createSecret(client, auth, cr);
+}
+
+export interface IAppConfig {
+  name: string;
+  namespace: string;
+  version: string;
+  chartName: string;
+  catalogName: string;
+  configMapContents: string;
+  secretContents: string;
+}
+
+/**
+ * Install an app in a cluster, as well as create
+ * the required ConfigMap and Secret resources to
+ * fully configure it.
+ * @param clientFactory
+ * @param auth
+ * @param clusterID
+ * @param appConfig
+ */
+export async function createApp(
+  clientFactory: HttpClientFactory,
+  auth: IOAuth2Provider,
+  clusterID: string,
+  appConfig: IAppConfig
+): Promise<applicationv1alpha1.IApp> {
+  const clusterConfigMapName = getClusterConfigMapName(
+    clusterID,
+    appConfig.chartName
+  );
+
+  const kubeConfigSecretName = getKubeConfigSecretName(clusterID);
+
+  const app: applicationv1alpha1.IApp = {
+    apiVersion: 'application.giantswarm.io/v1alpha1',
+    kind: 'App',
+    metadata: {
+      name: appConfig.name,
+      namespace: clusterID,
+      labels: {
+        [applicationv1alpha1.labelAppOperator]: '1.0.0',
```

Comment:
Are you sure this is a fixed version? I thought that depends on the release.

Reply:
We have this in `kubectl gs` and the REST API, as well. After talking to Batman, seems that the reason behind it is that apps running on clusters that have older release versions (with app-operator older than `v3.0.0`) don't go through the `app-admission-controller`, so they need this label pre-set.

More on that:
https://docs.giantswarm.io/app-platform/defaulting-validation/

Reply:
Sure, if Batman says so.